apm-agent-core/src/test/java/co/elastic/apm/agent/report/ApmServerClientTest.java

```diff
@@ -0,0 +1,148 @@
+    }
+
+    @Test
+    public void testUseNextUrlOnError() throws Exception {
+        apmServerClient.incrementAndGetErrorCount(0);
+        assertThat(apmServerClient.getCurrentUrl().getPort()).isEqualTo(apmServer2.port());
+
+        apmServerClient.execute("/test", HttpURLConnection::getResponseCode);
+
+        apmServer1.verify(0, getRequestedFor(urlEqualTo("/test")));
+        apmServer2.verify(getRequestedFor(urlEqualTo("/test")));
+    }
+
+    @Test
+    public void testWrapUrlsOnConsecutiveError() throws Exception {
+        int expectedErrorCount = apmServerClient.incrementAndGetErrorCount(0);
+        apmServerClient.incrementAndGetErrorCount(expectedErrorCount);
+
+        testInitialCurrentUrlIsFirstUrl();
+    }
+
+    @Test
+    public void testRetry() throws Exception {
+        assertThat(apmServerClient.<String>execute("/test", conn -> new String(conn.getInputStream().readAllBytes()))).isEqualTo("hello from server 2");
+        assertThat(apmServerClient.getCurrentUrl().getPort()).isEqualTo(apmServer2.port());
+        apmServer1.verify(getRequestedFor(urlEqualTo("/test")));
+        apmServer2.verify(getRequestedFor(urlEqualTo("/test")));
+    }
+
+    @Test
+    public void testRetryFailure() {
+        assertThatThrownBy(() -> apmServerClient.execute("/not-found", URLConnection::getInputStream))
+            .isInstanceOf(FileNotFoundException.class)
+            .matches(t -> t.getSuppressed().length == 1, "should have a suppressed exception");
+        apmServer1.verify(getRequestedFor(urlEqualTo("/not-found")));
+        apmServer2.verify(getRequestedFor(urlEqualTo("/not-found")));
+        // two failures -> urls wrap
+        assertThat(apmServerClient.getCurrentUrl().getPort()).isEqualTo(apmServer1.port());
+    }
+
+    @Test
+    public void testExecuteSuccessfullyForAllUrls() throws Exception {
+        apmServerClient.executeForAllUrls("/not-found", connection -> {
+            connection.getResponseCode();
+            return null;
+        });
+        apmServer1.verify(getRequestedFor(urlEqualTo("/not-found")));
+        apmServer2.verify(getRequestedFor(urlEqualTo("/not-found")));
+        // no failures -> urls in initial state
+        assertThat(apmServerClient.getCurrentUrl().getPort()).isEqualTo(apmServer1.port());
+    }
+
+    @Test
+    public void testExecuteFailureForAllUrls() {
+        assertThatThrownBy(() -> apmServerClient.executeForAllUrls("/not-found", connection -> {
+            connection.getInputStream();
+            return null;
+        }))
+            .isInstanceOf(FileNotFoundException.class)
+            .matches(t -> t.getSuppressed().length == 1, "should have a suppressed exception");
+        apmServer1.verify(getRequestedFor(urlEqualTo("/not-found")));
+        apmServer2.verify(getRequestedFor(urlEqualTo("/not-found")));
+        // two failures -> urls wrap
+        assertThat(apmServerClient.getCurrentUrl().getPort()).isEqualTo(apmServer1.port());
```

Comment:
If you agree `executeForAllUrls` shouldn't affect state, the related test should be testing failure URLs and verifying state didn't change